nexus/handle.go

```diff
@@ -0,0 +1,155 @@
+	"errors"
+	"fmt"
+	"net/http"
+	"time"
+)
+
+// An OperationHandle is used to cancel operations and get their result and status.
+type OperationHandle struct {
+	// Name of the Operation this handle represents.
+	Operation string
+	// Handler generated ID for this handle's operation.
+	ID     string
+	client *Client
+}
+
+// GetInfoOptions are options for [OperationHandle.GetInfo].
+type GetInfoOptions struct {
+	// Header to attach to the HTTP request. Optional.
+	Header http.Header
+}
+
+// GetInfo gets operation information, issuing a network request to the service handler.
+func (h *OperationHandle) GetInfo(ctx context.Context, options GetInfoOptions) (*OperationInfo, error) {
+	url, err := h.client.joinURL(h.Operation, h.ID)
+	if err != nil {
+		return nil, err
+	}
+	httpReq, err := http.NewRequestWithContext(ctx, "GET", url.String(), nil)
+	if err != nil {
+		return nil, err
+	}
+	if options.Header != nil {
+		httpReq.Header = options.Header.Clone()
+	}
+
+	httpReq.Header.Set(headerUserAgent, userAgent)
+	response, err := h.client.Options.HTTPCaller(httpReq)
+	if err != nil {
+		return nil, err
+	}
+
+	// Do this once here and make sure it doesn't leak.
+	body, err := readAndReplaceBody(response)
+	if err != nil {
+		return nil, err
+	}
+
+	if response.StatusCode != http.StatusOK {
+		return nil, newUnexpectedResponseError(fmt.Sprintf("unexpected response status: %q", response.Status), response, body)
+	}
+
+	return operationInfoFromResponse(response, body)
+}
+
+// GetResultOptions are Options for [OperationHandle.GetResult].
+type GetResultOptions struct {
+	// Header to attach to the HTTP request. Optional.
+	Header http.Header
+	// Duration to wait for operation completion. Zero or negative value implies no wait.
+	Wait time.Duration
+}
+
+// GetResult gets the result of an operation, issuing a network request to the service handler.
+//
+// By default, GetResult returns (nil, [ErrOperationStillRunning]) immediately after issuing a call if the operation has
+// not yet completed.
+//
+// Callers may set GetResultOptions.Wait to a value greater than 0 to alter this behavior, causing the client to long
+// poll for the result issuing one or more requests until the provided wait period exceeds, in which case (nil,
+// [ErrOperationStillRunning]) is returned.
+//
+// The wait time is capped to the deadline of the provided context.
+//
+// ⚠️ If a response is returned, its body must be read in its entirety and closed to free up the underlying connection.
+func (h *OperationHandle) GetResult(ctx context.Context, options GetResultOptions) (*http.Response, error) {
+	url, err := h.client.joinURL(h.Operation, h.ID, "result")
+	if err != nil {
+		return nil, err
+	}
+	httpReq, err := http.NewRequestWithContext(ctx, "GET", url.String(), nil)
+	if err != nil {
+		return nil, err
+	}
+	if options.Header != nil {
+		httpReq.Header = options.Header.Clone()
+	}
+	httpReq.Header.Set(headerUserAgent, userAgent)
+
+	startTime := time.Now()
+	for {
```

Comment:
This logic is a little funny and has a couple of problems.

First, you don't even guarantee one iteration if I set wait to 1ms (so if my monotonic system clock resolution sucks, I may never make a call)

Second, you are not differentiating server max-timeout-reached from wait-amount-reached. So you are potentially making an unnecessary extra call on server/client clock skew. For example, if I set wait to `10m` and the server returns still-running after `10m` but `wait - time.Since(startTime)` is like `2ms` (i.e. it hasn't reached exactly `10m` yet), you're gonna make another call. The server needs to return still-running _only_ on wait timeout (or unset wait). If the server times out _while_ waiting, that should be a different code and only on that code should you retry.

Reply:
> First, you don't even guarantee one iteration if I set wait to 1ms (so if my monotonic system clock resolution sucks, I may never make a call)

That's actually not a bad thing IMHO, it's unlikely that an HTTP request will return in less than a few single digit milliseconds.
Question is if the return code should be `ErrOperationStillRunning` in that case.

> If the server times out while waiting, that should be a different code and only on that code should you retry.

That would be an unexpected response error, the server is expected to return a 409 and that's the only code the client will retry.

Reply:
I think you'll like this version better:

```
	for {
		response, err := h.client.sendGetOperationResultRequest(ctx, httpReq, wait)
		if err != nil {
			wait = options.Wait - time.Since(startTime)
			if errors.Is(err, ErrOperationStillRunning) && wait > 0 {
				continue
			}
			return nil, err
		}
		return response, nil
	}
```

Reply:
Actually, I think it should be:

```go
	wait := options.Wait
	for {
		response, err := h.client.sendGetOperationResultRequest(ctx, httpReq, wait)
		if err != nil {
			if errors.Is(err, ErrServerTimeout) && wait > 0 {
				continue
			}
			return nil, err
		}
		wait = options.Wait - time.Since(startTime)
		return response, nil
	}
```

A server should only return `ErrOperationStillRunning` when wait hit, and we should return that to the user like any other error. Only when a server hits some timeout on their side that is not the full wait should we try again. This is the most sane approach.

Reply:
We only retry when server returns precondition failed (a.k.a still-running), so that's already there.
Setting wait to 0 like you did, will cause us to issue another trailing non long poll call to get the result, I don't think we want that.

Reply:
IMO we should never retry is-running. The server should only return is-running if client-provided wait time is reached. If the server says "I didn't finish your call or reach your wait time (i.e. I timed out)", then we should retry. This is a very clear way to make sure you don't make unnecessary calls.

>  Setting wait to 0 like you did, will cause us to issue another trailing non long poll call to get the result, I don't think we want that.

It was meant to issue one more immediate non-blocking call